### ISSUE TITLE
EC/NF many Q-curve labels are wrong

### ISSUE THREAD

**Issue by JohnCremona** [LMFDB/lmfdb#2718]
Of 661073 elliptic curves over number fields, about 1/3 have the q_curve flag set to True and many of these are wrong.  For example *all* the curves over a couple of cubic fields I just looked at are tagged as Q-curves, and most are not.

The problem might be that in most cases I have not yet determined whether the curves are Q-curves, but the 'q_curve' column is a Boolean and I don't know if  Null values are allowed.

On  www.lmfdb.org  (which as of now is using the old mongo database) we have 460382 curves with q_curve=False, same as on beta.lmfdb.org.  But on www.lmfdb.org only 34902 curves have q_curve=True so there must be 165,789 which are neither True nor False, and in the postgres conversion these seem to have picked up the True value.

In any case I am working on code which will determine the Q-curve status of all, but that will take a bit of time, so a stop-gap solution may be needed.

**Comment by AndrewVSutherland** [comment]
Null values are allowed on any postgres datatype, although from a database design perspective I think it is better to reserve null to mean "n/a" (not applicable), rather than "unknown".  For cases where the is a boolean value that is hard to compute but in principal always applies to some well defined set of objects (in this case, all elliptic curves over number fields I assume) I think it is better to use a smallint column that takes values -1,0,1 where -1 means false, 0 means unknown, and 1 means true.

As a quick fix, would it suffice to simply scan the mongo db table and create a file with pairs label:q_curve where q_curve it True, False, or null and then apply this to the postgres database?

If so, I can easily get this done for you today.

**Comment by JohnCremona** [comment]
As a quick fix that would be extremely helpful.  I ought to be able to do that myself but while I have now done a successful table rewrite (on the same table) I do not know how to change a column type at the same time.

It's also the case that I am not certain that the old values are all correct since I realised that I was using an incorrect criterion at some point, and curves uploaded after that point were tagged as unknown, but I don't think that I went back to rewrite what had already been uploaded.  That will of course all be fixed in due course, unless you wanted to take this opportunity to mark them *all* as unknown which at least will eliminate any erroneous tags.

**Comment by AndrewVSutherland** [comment]
Setting them all to null would be very easy.  Just drop the column and re-add it.  You could also change the type at the same time if you wish (it will have whatever type you specify when you re-add the column).  But you should update the code to handle -1,0,1 before setting these values because the code is liable to treat -1 as true if it is expecting a boolean.

One caveat is that you want to be sure the code won't fail on a null value before you drop the column, but I presume that is true since in the mongo db version it is often null.

**Comment by JohnCremona** [comment]
Right now the code does

>         # Q-curve / Base change
        self.qc = self.q_curve
        if self.qc == "?":
            self.qc = "not determined"
        elif self.qc == True:
            self.qc = "yes"
        elif self.qc == False:
            self.qc = "no"
        else: # just in case
            self.qc = "not determined"

So certainly code change will be required to handle values in {-1,0,1}.  I guess the thing to do here (for the web page display) would be to map 1 and True to "yes", 0 and False to "no" and anything else to "not determined".  That will make the display OK.  Then for the query, the only values ever searched for now are True and False, so that will need changing.  It seems that whatever I do, in whatever order,  something will be broken at some point.

**Comment by AndrewVSutherland** [comment]
Then why don't we just leave it as a boolean, I don't think it is necesary to switch to -1,0,1.  The easy fix would then be to simply drop and re-add the column (which will make it null everywhere), and then fill in values with true/false as you determine them.

**Comment by JohnCremona** [comment]
After #2721 is merged the code will no longer read the q_curve flags from the database at all, at which point we can easily change the data format.
I personally am OK with using Null to mean "not yet computed", rather than an arbitrary invalid value.

**Comment by JohnCremona** [comment]
I am currently processing all the curves over quadratic fields (which is quite a large proportion, and also the easiest case).  Compared with the old data (on www.lmfdb.org, i.e. mongo) there are quite a lot which were wrongly tagged as False (not Q-curves).  This is because the test I was applying only checked whether the conjugate curve was isogenous over the base field.  A correct test (valid over all Galois number fields) is that the set of j-invariants of the isogeny class is closed under conjugation, or equivalently if that set has a single complete Galois orbit.  (So a special case is: if there is a curvein the isogeny class with rational j-invariant then the curves in the class are Q-curves.)

That should finish today, and tomorrow I will upload these corrected values for all the curves over quadratic fields, which is 491051/661073 or about 74% of the total.

**Comment by JohnCremona** [comment]
I am now ready to rewrite the ec_nfcurves table with correct q_curve flags for all *quadratic* fields.  At the same time I would like to set all the flags for all curves over fields of degree >2 (i.e. 170022 curves) to Null, pending the correct calculation of these.  @AndrewVSutherland or @edgarcosta or @roed314 : if in my python dict for a curve C I set C['q_curve']=None, will that have the right effect?

**Comment by edgarcosta** [comment]
Yes

Alternatively, you can do it in SQL (which will be much faster)

```
UPDATE ec_nfcurves SET q_curve = NULL WHERE degree != 2
```

This is one command, but if you ever want to do multiple commands be able to rollback if something goes wrong you can do
```
BEGIN;
UPDATE ec_nfcurves SET q_curve = NULL WHERE degree != 2
etc etc
COMMIT;
```
see: https://www.postgresql.org/docs/10/tutorial-transactions.html

**Comment by JohnCremona** [comment]
Thanks @edgarcosta .  Since I am actually setting this to a genuine value for thousands of items (rows) in a rewrite, I'll do it that way.
For the faster SQL way,   an ignoramus like me needs more instructions.  I know that the SQL commands you gave (and those in the tutorial link) need to be given after to some interactive system,  but what exactly does one need to type *at the bash command line* to get the point where such a command is possible?  Adding such instructions to @roed314 's recently added SQL tips section to the FAQ would make that usable too.

**Comment by roed314** [comment]
Good idea; I'll add it now.

In the mean time, while logged into legendre, type
`psql -d lmfdb` and then give the lmfdb password.  This will give you read-only access to the database.

If you want write access (which would be required for what Edgar suggested), type
`psql -U editor -d lmfdb` and then give the editor password.

This will give you a prompt where you can type SQL commands.  Don't forget ending semicolons, which are not required when executing commands from python.

Alternatively, you can use the `db._execute` method.  If you use this route, make sure to wrap your statements in the `SQL` class from `psycopg2.sql` (you can also import it from `lmfdb.backend.database`).  You can see lots of examples of this paradigm in `lmfdb/backend/database.py`.

**Comment by JohnCremona** [comment]
Thanks @roed314 .   I should read a tutorial and try some things out in read-only mode -- it may be a while before I try any editing this way!

**Comment by roed314** [comment]
See Question 7 of Database interface at https://github.com/roed314/lmfdb/blob/master/Postgres_FAQ.md

We'll include this in our upcoming knowls PR.

**Comment by JohnCremona** [comment]
@roed314 I think that you have changed the ec_nfcurves table?  I now see that as well as 'heights' which is a list of LmfdbRealLiteral with column type numeric[] there is also old_heights which is a list of floats with column type jsonb.   Maybe it is unreasonable of me to expect that if a table schema changes then this should be recorded somewhere, and the editor responsible for this table be told?

Anyway, that should not have mattered but I just tried to do a rewrite on that table and it failed:

> /scratch/home/jcremona/lmfdb/lmfdb/backend/database.pyc in rewrite(self, func, query, resort, reindex, restat, tostr_func, commit, **kwds)
   2049                     for rec in self.search(query, projection=projection, sort=[]):
   2050                         processed = func(rec)
-> 2051                         searchfile.write(u'\t'.join(tostr_func(processed.get(col), self.col_type[col]) for col in search_cols) + u'\n')
   2052                         if self.extra_table is not None:
   2053                             extrafile.write(u'\t'.join(tostr_func(processed.get(col), self.col_type[col]) for col in extra_cols) + u'\n')

>/scratch/home/jcremona/lmfdb/lmfdb/backend/database.pyc in <genexpr>((col,))
   2049                     for rec in self.search(query, projection=projection, sort=[]):
   2050                         processed = func(rec)
-> 2051                         searchfile.write(u'\t'.join(tostr_func(processed.get(col), self.col_type[col]) for col in search_cols) + u'\n')
   2052                         if self.extra_table is not None:
   2053                             extrafile.write(u'\t'.join(tostr_func(processed.get(col), self.col_type[col]) for col in extra_cols) + u'\n')

>/scratch/home/jcremona/lmfdb/lmfdb/backend/encoding.pyc in copy_dumps(inp, typ)
    339         return r'\\x' + ''.join(c.encode('hex') for c in inp)
    340     else:
--> 341         raise TypeError("Invalid input %s (%s) for postgres type %s"%(inp, type(inp), typ))

>TypeError: Invalid input [0.0458094381868886] (<type 'list'>) for postgres type numeric[]

so something is broken and not by me!  This rewrite was *only* changing the q_curve column values.

I will assume that as the rewrite failed, the table was unchanged.

I would have liked to fix some incorrect data to address this critical issue but it will not now happen today.

**Comment by roed314** [comment]
Oops, I'm sorry!  After our discussion on heights last week I changed the type and ported over the data, but then got distracted by other work and forgot to write to you about it.  If the new height data looks okay then the old heigh column can be dropped.  I think it's fine but didn't want to drop it before running it by you.

I will look into the encoding issue.

**Comment by roed314** [comment]
See https://github.com/LMFDB/lmfdb/pull/2727

**Comment by JohnCremona** [comment]
@edgar It did *not* work to set the q_curve flag to None since then when you get a row from that table it has no q_curve attribute set at all.  This leads to current errors e.g. http://beta.lmfdb.org/EllipticCurve/3.3.1957.1/2.1/a/1

Is it documented anywhere what happens when you retrieve an object from the database where one field (column) is Null?  From what I see, the dictionary does not have that key at all.

I can fix this in the code, but note that the current master has effectively commented out all the Q-curve stuff while I recompute it all, so pushing master to dev is the simplest.

**Comment by JohnCremona** [comment]
OK, that is now fixed.

**Comment by roed314** [comment]
> Is it documented anywhere what happens when you retrieve an object from the database where one field (column) is Null? From what I see, the dictionary does not have that key at all.

This behavior is a legacy from Mongo, since Mongo didn't have a notion of columns and if a key wasn't present it didn't show up.  Some LMFDB code depended on this behavior.  We'd really like to switch to having the dictionary always include all keys requested, but haven't gotten around to making this happen yet.  In the mean time, None values are elided from the returned dictionary.

**Comment by JohnCremona** [comment]
Using the fix at #2727 I did a successful rewrite adding correct Q-curve flags to all curves over fields of degree 2.  Since then I computed all the others and right now am doing a rewrite adding all those too.  After that there should be no curves in ec_nfcurves with Null (or incorrect) Q-curve flag.  The I'll make a new PR to un-hide the stuff hidden at #2721 .  After that is done and merged this can be closed.
